client/admin.go

```diff
@@ -0,0 +1,190 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/base"
+	"github.com/cockroachdb/cockroach/util"
+)
+
+const (
+	// Types of admin config to manipulate.
+
+	// Accounting expects proto.AcctConfig
+	Accounting = "acct"
+	// Permission expects proto.PermConfig
+	Permission = "perms"
+	// Quit only handles Get requests.
+	Quit = "quit"
+	// User expects proto.UserConfig
+	User = "users"
+	// Zone expects proto.ZoneConfig.
+	Zone = "zones"
+)
+
+// AdminClient issues http requests to admin endpoints.
+// TODO(marc): unify the way we handle addresses in clients.
+type AdminClient struct {
+	context    *base.Context
+	address    string
+	configType string
+}
+
+// NewAdminClient returns a new 'configType' admin client, talking to 'address'.
+func NewAdminClient(ctx *base.Context, address, configType string) AdminClient {
+	return AdminClient{ctx, address, configType}
+}
+
+// adminURI builds a base URI for the embedded 'configType'
+func (a *AdminClient) adminURI() string {
+	return fmt.Sprintf("%s://%s/_admin/%s", a.context.RequestScheme(),
+		a.address, a.configType)
+}
+
+// adminURIWithKey builds a URI for the 'configType' and key.
+func (a *AdminClient) adminURIWithKey(key string) string {
+	return fmt.Sprintf("%s/%s", a.adminURI(), url.QueryEscape(key))
+}
+
+// Get issues a GET and returns the plain-text body. It cannot take a key.
+func (a *AdminClient) Get() (string, error) {
+	req, err := http.NewRequest("GET", a.adminURI(), nil)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Add(util.AcceptHeader, util.PlaintextContentType)
+	body, err := a.do(req)
+	if err != nil {
+		return "", err
+	}
+	return string(body), nil
+}
+
+// GetJSON issues a GET request and returns a json-encoded response.
+func (a *AdminClient) GetJSON(key string) (string, error) {
+	req, err := http.NewRequest("GET", a.adminURIWithKey(key), nil)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Add(util.AcceptHeader, util.JSONContentType)
+	body, err := a.do(req)
+	if err != nil {
+		return "", err
+	}
+	return string(body), nil
+}
+
+// GetYAML issues a GET request and returns a yaml-encoded response.
+func (a *AdminClient) GetYAML(key string) (string, error) {
+	req, err := http.NewRequest("GET", a.adminURIWithKey(key), nil)
```

Comment:
looks like all of the http traffic so far could just call out to a souped-up `do`: `func do(method, uri, header, accept, body string)` or something like that.